text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,236 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [amaranth-lang/rfcs#0036](https://github.com/amaranth-lang/rfcs/pull/0036)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Async testbench functions
+
+## Summary
+[summary]: #summary
+
+Introduce an improved simulator testbench interface using `async`/`await` style coroutines.
+
+## Motivation
+[motivation]: #motivation
+
+For the purpose of writing a testbench, an `async` function will read more naturally than a generator function, especially when calling subfunctions/methods.
+
+A more expressive way to specify trigger/wait conditions allows the condition checking to be offloaded to the simulator engine, only returning control to the testbench process when it has work to do.
+
+Passing a simulator context to the testbench function provides a convenient place to gather all simulator operations.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As an example, let's consider a simple stream interface with `valid`, `ready` and `data` members.
+On the interface class, we can then implement `.send()` and `.recv()` methods like this:
+
+```python
+class StreamInterface(PureInterface):
+    async def recv(self, sim):
```

Comment:
To be honeat I feel that as written this example is still fairly misleading as not only to capabilities of this RFC but also currently in-pipeline RFCs and it would be better to rewrite these methods as function (non-method) simulation helpers, to avoid any ambiguity.